File: ts/instructions/moveNote.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface MoveNoteAccounts {
  signer: PublicKey
  /** The forest */
  forest: PublicKey
  /** The tree */
  tree: PublicKey
  /** The node the note is currently attached to */
  sourceNode: PublicKey
  /** The node the note will be attached to */
  destinationNode: PublicKey
  /** The new note */
  note: PublicKey
  /** Common Solana programs */
  systemProgram: PublicKey
}

export function moveNote(accounts: MoveNoteAccounts) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.signer, isSigner: true, isWritable: true },
    { pubkey: accounts.forest, isSigner: false, isWritable: false },
    { pubkey: accounts.tree, isSigner: false, isWritable: false },
    { pubkey: accounts.sourceNode, isSigner: false, isWritable: true },
    { pubkey: accounts.destinationNode, isSigner: false, isWritable: false },
    { pubkey: accounts.note, isSigner: false, isWritable: true },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([241, 53, 131, 127, 195, 204, 16, 220])
  const data = identifier
  const ix = new TransactionInstruction({ keys, programId: PROGRAM_ID, data })
  return ix
}

File: programs/dippies_index_protocol/src/instructions/move_note.rs
use anchor_lang::prelude::*;

use crate::errors::DipErrors;
use crate::seeds::{FOREST_SEED, NODE_SEED, NOTE_SEED, TREE_SEED};
use crate::state::{Forest, Node, Note, Tree};

pub fn move_note(ctx: Context<MoveNote>) -> Result<()> {
    msg!("Moving a note");

    let source_node = &mut ctx.accounts.source_node;
    let note = &mut ctx.accounts.note;
    note.parent = ctx.accounts.destination_node.key();

    let note_position = source_node.notes.iter().position(|&n| n == note.key());
    if note_position.is_some() {
        // Remove the note from the source
        source_node.notes.remove(note_position.unwrap());
        source_node.stake -= note.stake;
    }

    Ok(())
}

#[derive(Accounts)]
pub struct MoveNote<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    /// The forest
    #[account(
        seeds = [
            FOREST_SEED.as_bytes(),
            &forest.id.to_bytes(),
        ],
        bump,
    )]
    pub forest: Account<'info, Forest>,

    /// The tree
    #[account(
        seeds = [
            TREE_SEED.as_bytes(),
            &forest.key().to_bytes(),
            &tree.title.as_ref(),
        ],
        bump,
    )]
    pub tree: Account<'info, Tree>,

    /// The node the note is currently attached to
    #[account(
        mut,
        seeds = [
            NODE_SEED.as_bytes(),
            &tree.key().to_bytes(),
            &source_node.parent.key().to_bytes(),
            &source_node.tags.last().unwrap().as_ref(),
        ],
        bump,
    )]
    pub source_node: Account<'info, Node>,

    /// The node the note will be attached to
    #[account(
        seeds = [
            NODE_SEED.as_bytes(),
            &tree.key().to_bytes(),
            &destination_node.parent.key().to_bytes(),
            &destination_node.tags.last().unwrap().as_ref(),
        ],
        bump,
        constraint = destination_node.tags.iter().all(|x| note.tags.contains(x)) @ DipErrors::TagsMismatch,
    )]
    pub destination_node: Account<'info, Node>,

    /// The new note
    #[account(
        mut,
        seeds = [
            NOTE_SEED.as_bytes(),
            &tree.key().to_bytes(),
            &note.id.to_bytes()
        ],
        bump,
        constraint = note.parent == source_node.key() @ DipErrors::InvalidNode,
    )]
    pub note: Account<'info, Note>,

    /// Common Solana programs
    pub system_program: Program<'info, System>,
}
